system/lib/wasmfs/file.h

```diff
@@ -8,37 +8,133 @@
+  }
+
+  template <class T> T* cast() {
+    static_assert(std::is_base_of<File, T>::value, "File is not a base of destination type T");
+    assert(int(kind) == int(T::SpecificFileKind));
+    return (T*)this;
+  }
+
+  template <class T> const T* cast() const {
+    static_assert(std::is_base_of<File, T>::value, "File is not a base of destination type T");
+    assert(int(kind) == int(T::SpecificFileKind));
+    return (const T*)this;
+  }
+
+  class Handle {
+
+    std::unique_lock<std::mutex> lock;
 
+  protected:
+    std::shared_ptr<File> file;
+
+  public:
+    Handle(std::shared_ptr<File> file) : file(file), lock(file->mutex) {}
+  };
+
+  Handle get() { return Handle(shared_from_this()); }
+
+protected:
+  File(FileKind kind) : kind(kind) {}
   // A mutex is needed for multiple accesses to the same file.
   std::mutex mutex;
 
+private:
+  // TODO: Add other File properties later.
+
+  FileKind kind;
+};
+
+class DataFile : public File {
+
   virtual __wasi_errno_t read(const uint8_t* buf, __wasi_size_t len) = 0;
   virtual __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) = 0;
 
 public:
-  virtual ~File() = default;
-  class Handle {
-    File& file;
-    std::unique_lock<std::mutex> lock;
+  enum { SpecificFileKind = File::DataFileKind };
+  DataFile() : File(File::DataFileKind) {}
+  virtual ~DataFile() = default;
+  class Handle : public File::Handle {
+
+    DataFile* getFile() { return file.get()->dynCast<DataFile>(); }
 
   public:
-    Handle(File& file) : file(file), lock(file.mutex) {}
+    Handle(std::shared_ptr<File> dataFile) : File::Handle(dataFile) {}
     Handle(Handle&&) = default;
 
-    __wasi_errno_t read(const uint8_t* buf, __wasi_size_t len) { return file.read(buf, len); }
-    __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) { return file.write(buf, len); }
+    __wasi_errno_t read(const uint8_t* buf, __wasi_size_t len) { return getFile()->read(buf, len); }
+    __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) {
+      return getFile()->write(buf, len);
+    }
+  };
+
+  Handle get() { return Handle(shared_from_this()); }
+};
+
+class Directory : public File {
+protected:
+  std::unordered_map<std::string, std::shared_ptr<File>> entries;
+
+public:
+  enum { SpecificFileKind = File::DirectoryKind };
+  Directory() : File(File::DirectoryKind) {}
+  class Handle : public File::Handle {
+    Directory* getFile() { return file.get()->dynCast<Directory>(); }
+
+  public:
+    Handle(std::shared_ptr<File> directory) : File::Handle(directory) {}
+
+    std::shared_ptr<File> getEntry(std::string pathName) {
+      if (getFile()->entries.find(pathName) == getFile()->entries.end()) {
+        return nullptr;
+      } else {
+        return getFile()->entries[pathName];
+      }
+    }
+
+    void setEntry(std::string pathName, std::shared_ptr<File> inserted) {
+      getFile()->entries[pathName] = inserted;
+    }
+
+    // For debugging, TODO: delete later.
```

Comment:
It would be good to have a debugging macro like `DEBUG_WASMFS` to hide this behind.